File: 8/sol.py
from collections import deque
from sys import stdin
from itertools import product
import math

grid = [[int(d) for d in line[:-1]] for line in stdin]

def next_largest(arr):
    res = [-1 for _ in range(len(arr))]
    s = []
    for i in range(len(arr)):
        while s and arr[s[-1]] <= arr[i]: 
            res[s.pop()] = i
        s.append(i)
    return res

def prev_largest(arr):
    res = [-1 for _ in range(len(arr))]
    s = []
    for i in range(len(arr) -1, -1, -1):
        while s and arr[s[-1]] <= arr[i]:
            res[s.pop()] = i
        s.append(i)
    return res
    
def replace(grid, ov, nv):
  for i in range(len(grid)):
    for j in range(len(grid[i])):
      if grid[i][j] == ov:
        grid[i][j] = nv
  return grid

def transpose(grid):
  return [list(row) for row in zip(*grid)]

def format(grid):
    return "\n".join("".join(map(lambda x: str(x).rjust(3), row)) for row in grid)

n, m = len(grid), len(grid[0])

l2r = [next_largest(row) for row in grid]
r2l = [prev_largest(row) for row in grid]

u2d = transpose([next_largest([grid[r][c] for r in range(n)]) for c in range(m)])
d2u = transpose([prev_largest([grid[r][c] for r in range(n)]) for c in range(m)])

def part1(l2r, r2l, u2d, d2u):
    res = 0
    for r, c in product(range(n), range(m)):
        res += -1 in [l2r[r][c], r2l[r][c], u2d[r][c], d2u[r][c]]
    return res

def part2(l2r, r2l, u2d, d2u):
    n, m = len(l2r), len(l2r[0])
    mx = -math.inf

    for r, c in product(range(n), range(m)):
        l2r[r][c] = m - 1 if l2r[r][c] == -1 else l2r[r][c]
        r2l[r][c] = 0 if r2l[r][c] == -1 else r2l[r][c]
        u2d[r][c] = n - 1 if u2d[r][c] == -1 else u2d[r][c]
        d2u[r][c] = 0 if d2u[r][c] == -1 else d2u[r][c]

        l2r_vd = l2r[r][c] - c
        r2l_vd = c - r2l[r][c]
        u2d_vd = u2d[r][c] - r
        d2u_vd = r - d2u[r][c]
        mx = max(mx, l2r_vd * r2l_vd * u2d_vd * d2u_vd)
    return mx

print(part1(l2r, r2l, u2d, d2u))
print(part2(l2r, r2l, u2d, d2u))

File: 8/sol.rs
use std::io::{self, BufRead};

// first larger after, depends on iter order
fn calc<I>(iter: I, size: usize) -> Vec<Option<isize>> 
where
    I: Iterator<Item = (usize, usize)>
{
    let mut res = vec![None; size];

    let mut s: Vec<(usize, usize)> = Vec::new();

    for (i, n) in iter {
        while !s.is_empty() && (n >= s.last().unwrap().1) {
            let j = s.pop().unwrap().0;
            res[j] = Some(i as isize);
        }
        s.push((i, n));
    }
    res
}

fn transpose<T: std::clone::Clone>(a: Vec<Vec<T>>) -> Vec<Vec<T>> {
    let mut res = vec![vec![]; a[0].len()];
    for i in 0..a.len() {
        for j in 0..a[i].len() {
            res[j].push(a[i][j].clone());
        }
    }
    res
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();

    let lines = stdin
        .lock()
        .lines()
        .collect::<io::Result<Vec<String>>>()?;

    let grid: Vec<Vec<usize>> = lines
        .iter()
        .map(|line|
             line
                .chars()
                .map(|c| c.to_digit(10).unwrap() as usize)
                .collect())
        .collect();

    
    let n = grid.len();
    let m = grid[0].len();

    let l2r: Vec<Vec<_>> = grid
        .iter()
        .map(|row| calc(
                row.iter().enumerate().map(|(a,b)| (a, *b)),
                n)
        )
        .collect();

    let r2l: Vec<Vec<_>> = grid
        .iter()
        .map(|row| calc(
                row.iter().enumerate().rev().map(|(a,b)| (a, *b)),
                n)
        )
        .collect();

    
    let u2d: Vec<Vec<_>> = transpose((0..m)
        .map(|i| calc(
                (0..n).map(|j| (j, grid[j][i])),
                m)
        )
        .collect());

    let d2u: Vec<Vec<_>> = transpose((0..m)
        .map(|i| calc(
                (0..n).map(|j| (j, grid[j][i])).rev(),
                m)
        )
        .collect());

    let p1: usize = (0..n)
        .map(|i| (0..m)
             .map(|j| {
                (l2r[i][j].is_none() ||
                 r2l[i][j].is_none() || 
                 u2d[i][j].is_none() ||
                 d2u[i][j].is_none()) as usize
             })
             .sum::<usize>()
        )
        .sum();

    let p2 = (0..n)
        .map(|row| (0..m)
             .map(|col| {
                let l2ri = l2r[row][col].unwrap_or((n - 1) as isize);
                let r2li = r2l[row][col].unwrap_or(0);
                let u2di = u2d[row][col].unwrap_or((m - 1) as isize);
                let d2ui = d2u[row][col].unwrap_or(0);
                let l = l2ri - (col as isize);
                let r = (col as isize) - r2li;
                let u = u2di - (row as isize);
                let d = (row as isize) - d2ui;
                l * r * u * d
            })
            .max()
            .unwrap()
        )
        .max()
        .unwrap();

    println!("part 1: {}", p1);
    println!("part 2: {}", p2);

    Ok(())
}
